lib/gateway.js

```diff
@@ -281,6 +288,47 @@ async function buildGateway (gatewayOpts, app) {
     schema,
     serviceMap,
     entityResolversFactory: factory,
+    pollingInterval: gatewayOpts.pollingInterval,
+    async refresh () {
+      if (this._serviceSDLs === undefined) {
+        this._serviceSDLs = serviceSDLs.join('')
+      }
+
+      await allSettled(
+        Object.values(serviceMap).map((service) =>
+          service.refresh().catch((err) => {
+            if (service.mandatory) {
+              throw Object.assign(err, service)
+            }
+          })
+        )
+      ).then((result) => {
+        result
+          .filter(({ reason }) => reason !== undefined)
+          .forEach(({ reason }) => errorHandler(reason, reason.service))
+      })
```

Comment:
I think if some of the services are offline we should not start the gateway.

Reply:
@mcollina That's why I asked for the `mandatory` flag. In some scenarios it might be that the network has a temprorary glitch, or some of the services (which might not be under the control of the gateway development team) eventually start after the Gateway has already been booted.

We fac ethis issue, as we create a new integration test environment for each of our projects merge requests; the Java based services of another team (which are part of the federation) usually take longer to boot than our services; so we integrate them after the Gateway has already been started.
Their functionality is not mandatory for the system basics to work, so it's fine if they're not present.

F.e. we work on a Smart Home product, the core functionality is steering devices directly & execute "if this than that" like rules. We also have additional features, like historic data of device states; this isn't considered mandatory. It's "fine" if that feature, which is also part of the federation, is temporarily down, as long as the main functionality works.

this way, we can have a resiliant system which makes sure that the main functionality of the system is working fine & while it's temporarily okay for 2nd level systems, with non critical functionailty, to fail.

Reply:
but, what if offline service is not mandatory or service is restarting right now?

for example, you have the next services:
* User-Service (mandatory)
* Company-Service (not mandatory)
* Gateway Service (core)

You're developing Company-Service and it's restarting. Gateway Service at the same time tries to fetch new schema by using `pollingInterval`. Company-Service can't answer because it's restarting. Should we shutdown Gateway-Service in this case? After some time, Company-Service will be ready to work, so Gateway-Service will fetch correct schema and can continue working with it.

I guess this [solution](https://github.com/mcollina/fastify-gql/pull/206#discussion_r461401701) will be best. In this case, the Developer can operate with any error by himself.

What do you think?

Reply:
I also think that allowing services "hot-plug" to the gateway would make it a lot more flexible for this kind of environment ! But what makes a service "mandatory" or "not mandatory" ?

Reply:
Having a `mandatory` flag is good.

Reply:
Good. Now we have it! :+1: